[FEATURE ember-container-inject-owner] Inject fake container with deprecations.

dgeb opened the pull request:
The fake container provides access to the following methods:
- `lookup`
- `lookupFactory`

These methods are mapped through the equivalent `ContainerProxy` methods
(`lookup` and `_lookupFactory`) on the container's owner. 

Deprecation warnings are provided.

TBD: deprecation id and url

rwjblue commented:
Let's use the id and URL that was in the deprecation being removed here (I have already linked that url up with a decent deprecation guide entry).

rwjblue commented:
Also, I think we should leave the current deprecation (on property getter for 'container'). The fake container will only be used during instantiation of objects that do not have an .extend (as all of those that do have already had the getter added).

dgeb commented:
Sounds good, @rwjblue. I'll make those changes.

Would you also prefer that the FakeContainer not proxy through to the owner's equivalent methods, but instead just proxy through to the container's methods? Otherwise accessing `container` at different times could theoretically produce different behavior. Not to mention that FakeContainer -> owner -> container is pretty circuitous.

rwjblue commented:
> Would you also prefer that the FakeContainer not proxy through to the owner's equivalent methods, but instead just proxy through to the container's methods?

Yes, but if the `FakeContainer` is essentially only exposed externally by [this branch](https://github.com/emberjs/ember.js/blob/0810d913f9d358e5d37d25a1efd2295ebaebc6e2/packages/container/lib/container.js#L333).

---

I think the changes to the PR would be (only noting them because I just traced the various pathways):

* Remove [this conditional](https://github.com/emberjs/ember.js/blob/0810d913f9d358e5d37d25a1efd2295ebaebc6e2/packages/container/lib/container.js#L278-L280) and instead always add the `FakeContainer` there in `injectionsFor`.
* For any factories with `.extend` we will override that `FakeContainer` injection with the current property getter deprecation [here](https://github.com/emberjs/ember.js/blob/0810d913f9d358e5d37d25a1efd2295ebaebc6e2/packages/container/lib/container.js#L246-L248) (this is good)
* For any factories without `.extend` (or models with `MODEL_FACTORY_INJECTIONS` off) you would still get the `FakeContainer` due to the usage of `injectionsFor` in that [instantiate](https://github.com/emberjs/ember.js/blob/0810d913f9d358e5d37d25a1efd2295ebaebc6e2/packages/container/lib/container.js#L333)

dgeb committed:
[FEATURE ember-container-inject-owner] Inject fake container for non-extendable factories.

The fake container is required to provide access to the `container`
during instantiation of objects by non-extendable factories.

The fake container provides access to the following methods:

* `lookup`
* `lookupFactory`

Deprecation warnings are provided that direct the user to call
`getOwner` followed by the equivalent `ContainerProxy` methods
(`lookup` and `_lookupFactory`) on the owner.

rwjblue commented on packages/container/tests/container_test.js:
```diff
@@ -535,6 +535,39 @@ if (isEnabled('ember-container-inject-owner')) {
   });
+
+  QUnit.test('A deprecated `container` property is appended to every object instantiated from a non-extendable factory, and a fake container is available during instantiation.', function() {
+    let registry = new Registry();
```
Can you add `expect(5)` here (just to ensure that all the expectations are firing)?

rwjblue commented on packages/container/tests/container_test.js:
```diff
@@ -535,6 +535,39 @@ if (isEnabled('ember-container-inject-owner')) {
+        options.container.lookupFactory('abc:def');
+      }, 'Using the injected `container` is deprecated. Please use the `getOwner` helper to access the owner of this object and then call `_lookupFactory` instead.');
+
+      return new PostController(options);
```
Can you add:

```
// non-deprecated usage
let owner = getOwner(options)
owner.lookup('abc:def');
owner._lookupFactory('abc:def');
```

dgeb commented:
@rwjblue ok, just implemented your suggestions
